Reliability part2

Manghi opened the pull request:
I like keeping these PR's small so here's another. This one is just to keep things structured nice, tidy, and reusable.

Part two primarily includes:
- Create types for netwayste packet buffering
  - `ItemQueue<T>` is a `VecDeque<T>`
- Traits~!
  - `NetworkQueue<T>` describes the basic functionality needed for a transmit or receive network queue
  - `TXQueue<T>` implements `NetworkQueue<T>` for transmission-related actions
   - `RXQueue<T>` implements `NetworkQueue<T>` for receiving-related actions
- RX queues can work for any type T, as long as T is bound to `Sequenced`. 
  - This means we now have `RXQueue<Packet>` and `RXQueue<BroadcastChatMessages>` and the handling of them uses the exact same infrastructure!! (Yay no code duplication which was where I was headed at first)
- `Sequenced` is a trait which adds ordering to custom types, by comparing internal sequence numbers
  - For packets, this is their packet sequence number
  - For Updates, this would be the chat messages' sequence number
  - This is the crux in which ordering is determined when searching through a buffer for the insertion index
- General refactoring to ensure the surrounding code/comments actually make sense
- Few more unit tests

Manghi committed:
Buffer Response packets on tx (for server)

Manghi committed:
Implement out-of-order received packet buffering

- Maintains sorted VecDequeue by packet sequence number
- Does not yet handle wrapped sequence numbers (coming up next)
- Uses binary search on slices to determine where to insert in middle of the queue
- Added various buffer_rx_packet() unit tests (which surprisingly all passed. yay.)
- Also simplified logic in TX buffer pathway

Manghi committed:
(Does Not Pass All Tests) Adding out-of-order and wrapped rx packet buffering

Network Manager keeps track of sequence number wrapping point
Last two test cases need to be investigated:
- Wrapped value arrives before max sequence number
- Oldest packet arrives last

Manghi committed:
Addressed wrapping case when older/newer sequence numbers arrive before/after a wrap

Manghi committed:
Fixed more cases in wrapping seq number; more tests

Manghi committed:
Primarily cleaning up/adding comments

Manghi committed:
Some code clean up; Tried to clarify what the heck its doing

Manghi committed:
Refactored client chat handling to use if-let pattern

Manghi committed:
sequence_number() updated for Packet::Update::UniUpdateType

Manghi committed:
[DNC] Getting traits for generic tx/rx queues up :D

Manghi committed:
Completed NetworkQueue as a trait, see details

- Added a few shadow helper functions (like push_back) to `NetworkQueue`
to keep things reading cleanly.
- Updated all `self` references to use new structures
- Updated unit tests to reflect new impl of traits (all pass!)

Manghi committed:
Trying to make NetworkQueue generic

- Desire is to make it usable for packets or chat messages or anything
else sequenced and transfered over the network
- Added new Sequenced trait to help with the ordering of these
- Implemented NetworkQueue<T> for T=Packet; Next is BroadcastChatMessage

Manghi committed:
impl NetworkQueue<BroadcastChatMessage> for RXQueue<BroadcastChatMessage>

Also renamed buffer_packet to buffer_item to sound more generic

Manghi committed:
Client-side chat message buffering

Manghi committed:
Merge branch 'master' into reliability_part2

Manghi committed:
Generalized RXQueue<T> finally

- It makes sense to just pass in the type T  instead of throwing around
a sequence number.

Manghi committed:
Refactored NetworkQueue trait ecosystem to describe Items, not Packets

Manghi committed:
Use public BroadcastChatMessage::new()

AaronM04 commented:
@Manghi It looks like ticks happen 100x more often than they need to, and 99% of the time nothing gets run:

```rust
if (server_state.tick % 100) == 0
```
Maybe we could just remove that `if` block and decrease the frequency of the ticks?

Manghi commented:
> 
> 
> @Manghi It looks like ticks happen 100x more often than they need to, and 99% of the time nothing gets run:
> 
> ```rust
> if (server_state.tick % 100) == 0
> ```
> 
> Maybe we could just remove that `if` block and decrease the frequency of the ticks?

Yeah I'm aware of that, same thing happens in `client.rs`. Intention was to address them both in part three when I add in another future stream. Up to you if you want me to fix it in this PR or part 3.

Manghi committed:
Clear out queues on reset

AaronM04 commented on src/net.rs:
```diff
@@ -243,75 +332,381 @@ impl NetworkStatistics {
+    /// Search within the RX queue, but we have no idea where to insert.
+    /// This should cover only within the RX queue and not at the edges (front or back).
+    /// We accomplish this by splitting the VecDequeue into a slice tuple and then binary searching on each slice.
+    /// Small note: The splitting of VecDequeue is into its 'front' and 'back' halves, based on how 'push_front' and 'push_back' were used.
```
This is a very long line. Please split up!

AaronM04 commented on src/net.rs:
```diff
@@ -243,75 +332,381 @@ impl NetworkStatistics {
+    ///                     ^ wrapping index
+    ///
+    /// Because we can receive `T`'s out-of-order even when wrapped, there are checks added below to safeguard against this.
+    /// Primarily, they cover the cases where out-of-order insertion would transition the queue into a wrapped state from a non-wrapped state.
```
2 very long lines

AaronM04 commented on src/net.rs:
```diff
@@ -243,75 +332,381 @@ impl NetworkStatistics {
+
+    /// Checks if the insertion of `sequence` induces a newly wrapped queue state.
+    /// Sequence must be >=, or <, what we're comparing against. Cannot have wrapped yet.
+    fn is_seq_about_to_wrap(&self, buffer_wrap_index: Option<usize>, sequence: u64, oldest_seq_num: u64, newest_seq_num: u64) -> bool {
```
Please wrap this long line...

AaronM04 replied:
Also please rename to `will_seq_cause_a_wrap`! :smile:

AaronM04 commented on src/net.rs:
```diff
@@ -244,72 +333,391 @@ impl NetworkStatistics {
+    }
+}
+*/
+pub trait NetworkQueue<T: Ord+Sequenced> {
```
:skull: does not compile! Error on this line because `Sequenced` is not defined.
Rest of this review will omit latest commit. :skull_and_crossbones:

AaronM04 commented on src/net.rs:
```diff
@@ -244,72 +333,391 @@ impl NetworkStatistics {
+    /// have a corresponding sequence number to delineate order.
+    ///
+    /// This also handles the case where the sequence number numerically wraps.
+    /// `rx_buffer_wrap_index` is maintained to denote the queue index at which the numerical wrap occurs.
```
`rx_buffer_wrap_index` isn't defined anywhere. I think you meant `buffer_wrap_index`

AaronM04 commented on src/net.rs:
```diff
@@ -244,72 +333,391 @@ impl NetworkStatistics {
+    }
+
+    /// I've deemed 'far away' to mean the half of the max value of the type.
+    fn is_seq_sufficiently_far_away(&self, a: u64, b: u64) -> bool {
```
I think it would be cleaner to avoid requiring that `a >= b`. Why not have `mut a: u64` and `mut b: u64` in the argument list and then swap them if `a < b`? Then you can simplify `is_seq_about_to_wrap` and other usages of this function. :smile:

Manghi replied:
I don't have or anticipate a use case where I'd need them to be `mut` but I like your suggestion otherwise.

AaronM04 commented on src/net.rs:
```diff
@@ -244,72 +333,391 @@ impl NetworkStatistics {
+    }
+
+    /// Checks if the insertion of `sequence` induces a newly wrapped queue state.
+    /// Sequence must be >=, or <, what we're comparing against. Cannot have wrapped yet.
```
`Sequence must be >=, or <, what we're comparing against.`
Aren't those the only two possibilities? :smile:

Manghi replied:
😓 lol there's a secret third possibility which exists in our imagination. Clarified comment in next commit.

AaronM04 commented on src/net.rs:
```diff
@@ -244,72 +333,391 @@ impl NetworkStatistics {
+                    self.push_back(item);
+
+                    if let Some(buffer_wrap_index) = self.buffer_wrap_index {
+                        self.buffer_wrap_index = Some(buffer_wrap_index - 1);
```
Case is not hit by any unit test.

AaronM04 commented on src/net.rs:
```diff
@@ -244,72 +333,391 @@ impl NetworkStatistics {
+                self.insert_into_rx_queue(insertion_index, item);
+            } else {
+                // Smallest sequence number (in value) that we have seen thus far.
+                self.push_front(item);
```
Probably dead code. Please add `panic!` :boom:

AaronM04 commented on src/net.rs:
```diff
@@ -244,72 +333,391 @@ impl NetworkStatistics {
+                }
+            }
+        } else {
+            let insertion_index: Option<usize>;
```
Please comment because it's super far away from the `if`:
```rust
// sequence >= oldest_seq_num
```

AaronM04 commented:
Unit test names are wack bruh

Manghi committed:
Undoing Ord experiment so it now compiles...

Manghi committed:
Max line length of 120

Manghi committed:
Fixing wack names for unit tests and the other one

rename is_seq_about_to_wrap() to will_seq_cause_a_wrap()

Manghi committed:
Addressed comments clarity and simplified sequence wrapping checks

Manghi committed:
Fixes to buffer_item(), see notes below.

- Removed dead code in special case
- Added panic for probable dead code
- Added comment to remind us what else is handling

AaronM04 commented on src/net.rs:
```diff
@@ -401,20 +401,23 @@ pub trait NetworkQueue<T: Ord+Sequenced> {
-                && self.is_seq_sufficiently_far_away(newest_seq_num, sequence)
-            }
+            self.is_seq_sufficiently_far_away(sequence, oldest_seq_num)
+            && self.is_seq_sufficiently_far_away(sequence, newest_seq_num)
```
so much simpler :100:

AaronM04 reviewed (approved):
Looks good! Merging
